pkg/root/trusted_root.go

```diff
@@ -287,6 +275,68 @@ func ParseCertificateAuthority(certAuthority *prototrustroot.CertificateAuthorit
 	return certificateAuthority, nil
 }
 
+func ParseTimestampingAuthorities(certAuthorities []*prototrustroot.CertificateAuthority) (timestampingAuthorities []TimestampingAuthority, err error) {
+	timestampingAuthorities = make([]TimestampingAuthority, len(certAuthorities))
+	for i, certAuthority := range certAuthorities {
+		timestampingAuthority, err := ParseTimestampingAuthority(certAuthority)
+		if err != nil {
+			return nil, err
+		}
+		timestampingAuthorities[i] = timestampingAuthority
+	}
+	return timestampingAuthorities, nil
+}
+
+func ParseTimestampingAuthority(certAuthority *prototrustroot.CertificateAuthority) (TimestampingAuthority, error) {
+	if certAuthority == nil {
+		return nil, fmt.Errorf("CertificateAuthority is nil")
+	}
+	certChain := certAuthority.GetCertChain()
+	if certChain == nil {
+		return nil, fmt.Errorf("CertificateAuthority missing cert chain")
+	}
+	chainLen := len(certChain.GetCertificates())
+	if chainLen < 1 {
+		return nil, fmt.Errorf("CertificateAuthority cert chain is empty")
+	}
+
+	timestampingAuthority := &SigstoreTimestampingAuthority{
+		URI: certAuthority.Uri,
+	}
+	for i, cert := range certChain.GetCertificates() {
+		parsedCert, err := x509.ParseCertificate(cert.RawBytes)
+		if err != nil {
+			return nil, err
+		}
+		switch {
+		case i == 0 && !parsedCert.IsCA:
+			timestampingAuthority.Leaf = parsedCert
+		case i < chainLen-1:
+			timestampingAuthority.Intermediates = append(timestampingAuthority.Intermediates, parsedCert)
+		case i == chainLen-1:
+			timestampingAuthority.Root = parsedCert
+		}
+	}
+	validFor := certAuthority.GetValidFor()
+	if validFor != nil {
+		start := validFor.GetStart()
+		if start != nil {
+			timestampingAuthority.ValidityPeriodStart = start.AsTime()
+		}
+		end := validFor.GetEnd()
+		if end != nil {
+			timestampingAuthority.ValidityPeriodEnd = end.AsTime()
+		}
+	}
+
+	timestampingAuthority.URI = certAuthority.Uri
+
+	// TODO: Should we inspect/enforce ca.Subject?
```

Comment:
Can we either file issues for these TODOs or decide if these are needed?

My two cents:

* Subject is more of a human-readable value and not something we need to check. It's also the subject of the root, so there's not really anything to verify, as we implicitly trust the root.
* What is there to handle with the validity period? It seems like the above is reading it.

Reply:
I agree with you on both points! The validity period comment is out of date - I think that was added prior to moving the validity check into this function.